### DISCUSSION TITLE
Support für mehrere MPPTs

### DISCUSSION THREAD

**Discussion by FSys77** [hoylabs/OpenDTU-OnBattery#129] [Ideas]
Wäre es möglich, dass Du den Code so erweiterst, dass mehrere Victron MPPTs unterstützt werden?

Ich habe 2 Victrons. Der ESP32 hat ja afaik 3 x UART (und "beliebig" viele Softserials :-D).

**Comment by helgeerbe** [comment]
Hi @FSys77,

prinzipiell ist das möglich.  Allerdings müsste eine entsprechende Verwaltung implementiert werden. Freiwillige vor ;)
UARTS sind eigentlich keine mehr frei. Außer man verwendet das Pylontech Can Interface nicht. Die Debugschnittstelle würde ich nicht opfern wollen.

Wie viel Leistung die Nutzung von Softseraial benötigt, kann ich nicht sagen. Das müsste man dann mal ausprobieren.

**Reply by berni2288** [reply]
Alle Geräte wie Displays, CAN usw. sind eh optional, also das sollte kein Argument sein nicht mehrere Victrons zu unterstützen.
Ich denke das muss eh jeder OpenDTU-Benutzer sich selbst ausschnapsen wie er die UARTS & Pins verwendet, da sollte man keine Vorgaben machen. Das einzige was wirklich kritisch und required ist, ist die nrf24l01+ Anbindung.

**Comment by helgeerbe** [comment]
@FSys77 

Ich habe mir das noch mal durch den Kopf gehen lassen. 
Folgende Änderungen sehe ich:
- Serial device Manager
  - Liste von ca. 6 (?) Einträgen. Die ersten beiden wären hardware UART. Jeder Eintrag bekommt eine ID. Dort werden die TX, RX Pins gesetzte.
- Änderung der Can und Victron Handler, damit sie sie IDs verwenden
- Änderung der Can und Victron Handler, damit sie auch Software Serial nutzen
- Änderung des Victron Handlers, damit er mehrere Victrons auslesen kann
- Änderung Victron GUIs, damit man bis zu 5 Victrons auswählen kann
- Änderung der Victron MQTT damit auch alle Werte der MPPTs gesendet werden
- Änderung des Live Views analog Inverter zur Anzeige aller MPPTs
- Änderung von PowerLimiter, dass er von allen MPPTs die Panel Leistung für SolarPassThrough nimmt.
- Änderung des PinMappings 

Einschränkungen:
- x MPPTs laden **eine** Batterie
- die Batterie wird über **einen** Inverter entladen

Offene Punkte:
- Reicht die Sketch Size?
- Ist der esp32 performant genug für die Anzahl der Software Serial Schnittstellen?

**Reply by Adminius** [reply]
eine ernsthafte Frage: ist es nicht zu viel für kleinen ESP32? :)
Ich meine, 5 Victrons? bei den meistens Modulen braucht man schon 2 PV Module pro Victron um die benötigte Spannung zu erreichen.
Da wären wir schon bei 10 Modulen, sagen wir mal 4000 Wp.
Da ist auch 24V wohl nicht mehr das richtige, sondern eher 48V, wie z.B. Pylnontech.
Da reden wir auch nicht mehr über einen 300W WR, sondern her 1200 W damit das Ganze Sinn macht...

Das ist kein "Bahnkraftwerk" mehr, sondern schon eine richtig teurere und große Anlage.
Da würde es schon Sinn machen auf etwas größeres, wie RasPi zu gehen um das ganze schnell genug verarbeiten zu können.
Oder gleich über einen fertigen WR nachdenken (je nach Situation kann es sogar günstiger werden. 5x Victrons kosten auch viel Geld)
Max. 3 Victrons würden aus meiner Sicht Sinn machen. Gut, evtl. zu O/S/W noch Nord dazu, also 4x Victrons.

mit der Batterie will man ja auch noch reden können: CAN oder RS485/UART
Stromzähler je nach Typ evtl. auch noch RS485/UART (da bin ich schon dabei auf SW-Serial umzustellen)

so sind wir schon bei 6 UARTs. Wobei BMS würde ich auf HW UART wegen Datenrate hängen.

Zu den Einschränkungen: 
alle MPPTs laden eine Batterie ist aus meiner Sicht vollkommen in Ordnung
Warum muss aber die Batterie nur über einen Inverter entladen werden? Ich denke da an z.B. 3x 600W für 3 ph.

Bzgl. Performance: es wird wohl langsam Zeit beide Cores besser zu nutzen (wohl wissend, dass die single Core ESP32 ausgeschlossen werden)

**Reply by helgeerbe** [reply]
> eine ernsthafte Frage: ist es nicht zu viel für kleinen ESP32? :)

Bin im Moment ganz erstaunt was mit dem Ding so geht. Hätte eigentlich auch gedacht, die Loop wird langsam. Aber im Moment packt er es noch ganz gut. Memory haben wir auch wieder, nachdem die Partition angepasst worden ist. Aber ich habe tatsächlich keine Ahnung was Software Serial an Performance kostet. Muss ich noch testen.

Für mich persönlich ist allerdings nach einem weiteren MPPT auch Schluss. Der würde noch ganz gut passen, da meine Battery (2kWH) noch etwas Kappa für die Nacht hätte und die Investition (2 Panels und ein Lader) sich in Grenzen halten. Ansonsten würde ich tatsächlich auch eher eine große Anlage nehmen. 

Falls es mich überkommt kann ich noch einen weiteren Lader, 4 Panels und eine weitere Batterie hinzufügen. Aber dann habe ich schon 10 Panels und wüsste nicht, was ich mit dem Strom machen soll (wenn man tatsächlich nur 600w max aber konstant nutzen würde und könnte).

**Reply by Tyberyus** [reply]
Mal ne blöde Frage kann man nich Bluetooth verwenden um mit den Victrons zu kommunizieren? Hab mich noch nicht mit dem Aufbau und den Funktionen der Victron Bluetooth Integration beschäftigt könnte ich mir aber mal anschauen was da möglich wäre? Dann wäre noch ein UART frei.

**Reply by helgeerbe** [reply]
> Mal ne blöde Frage kann man nich Bluetooth verwenden um mit den Victrons zu kommunizieren?

Meines Wissens ist die Bluetooth Implementierung von Victron nicht offengelegt. Da müsste also jemand reverse Engineering machen. Im Gegensatz dazu hat Victron das Ve.Direct Protokoll offen gelegt. Aber wenn du da was findest, immer sehr gerne.

**Reply by berni2288** [reply]
Ist tauch die Frage wie gut dass funktionieren würde, wenn der ESP32 ständig zwischen WLAN und Bluetooth wechseln muss um Daten abzufragen.

**Reply by Tyberyus** [reply]
https://github.com/Fabian-Schmidt/esphome-victron_ble
Und https://github.com/keshavdv/victron-ble
das hab ich jetzt auf die Schnelle gefunden. Sieht nicht schlecht aus zumindest gibt es schon ein paar Sachen. Aber bei weitem noch nicht das was ich mir erhofft hatte.

**Reply by helgeerbe** [reply]
Danke für die Links. Ich werde damit mal heute Abend rumspielen, wäre cool, wenn das ginge.

**Comment by helgeerbe** [comment]
So, 

habe mal ein wenig experimentiert. War mir nicht ganz sicher ob Bluetooth und Wifi zusammen funktionieren, da sie sich die Antenne teilen. Aber ich konnte zumindest schon mal die Devices Scannen und danach war die OpenDTU wieder per WiFi erreichbar. Man sieht im Log, dass Wifi disconnected wird. Wie schnell es dann wieder geht, wenn Daten gelesen werden, werde ich noch ausprobieren.

````log
5:36:35.838 > Initialize Victron BLE interface... 
15:36:35.841 > BLE Scaning...
15:36:35.851 > WiFi disconnected
15:36:37.441 > Advertised Device: Name: SmartSolar XXXX, Address: e3:34:b8:98:26:8e, manufacturer data: e102100256a0017b05df48f20477cc6d65f4c1a532be 
15:36:37.484 > Advertised Device: Name: , Address: f1:bf:de:bb:3b:f3, manufacturer data: e1020292672fef98ec49b12bf18ef7aa5083c7520e711fd5e2b8c939cc 
15:36:37.816 > Advertised Device: Name: BatteryProtect XXXX, Address: ea:d1:19:15:16:3e, manufacturer data: e1021000b0a3, serviceUUID: cd54a3b0-880b-425b-b167-81ed6a15e913 
````
Manufacturer Data:
- startet mit MANUFACTURER_ID = 0x02E1 für Victron Energy BV.
- Byte 5 und 6 scheint die PID zu sein: 
  - 0XA056 = SmartSolar MPPT 100|30
 
Die anderen beiden Geräte haben kein Ve.Direct interface. Daher kenne ich die PID nicht.

Da ich jetzt die Adresse kenne, werde ich mal versuchen Kontakt aufzunehmen.

**Reply by helgeerbe** [reply]
Victron hat leider für die Solar Charger (im Moment) nur die advertising Infos veröffentlicht (https://community.victronenergy.com/questions/187303/victron-bluetooth-advertising-protocol.html)

Damit kann aktuell über Bluetooth nur ein reduzierter Satz an Infos geholt werden:

![image](https://user-images.githubusercontent.com/59169507/233853890-478301ff-60de-46c2-ade7-a0e3d181c4ad.png)


Ich werde das aus Spieltrieb mal implementieren und schauen, ob OpenDTU generell mit Bluetooth klar kommt. Hoffe aber, dass Victron auch demnächst den vollständigen Zugriff auf die Daten freigibt.

Außerdem muss man aktuell noch den encryption Key aus der App extrahieren. Victron plant allerdings ihn in der Smartphone App anzeigbar zu machen. So ist das für einfache User nicht ganz trivial.

**Reply by FSys77** [reply]
Wie machst Du das eigentlich mit dem Spielbetrieb? Trennst Du den Victron MPPT da vom System?

Hintergrund meiner Frage: Ich würde nämlich gerne auch mal damit herumprobieren; würde es funktionieren den an einen 12-V Bleiakku zu hängen, zum experiementieren? Würde der dann starten?

**Comment by swingstate** [comment]
Ich nehme an die Änderung von Parametern wie in der IOS App sind damit auch weiterhin nicht möglich - Beispielsweise das Ändern des Ladestroms.

**Reply by helgeerbe** [reply]
Nein, leider nicht. 

Laut https://community.victronenergy.com/questions/93919/victron-bluetooth-ble-protocol-publication.html stellen die Victrons einen weiteren Bluetooth Service zur Verfügung, der allerdings keine Änderungen erlaubt. Aktuell wird auch nur der Smart Shunt unterstützt. Victron selbst benutzt weiterhin ihren eignen Service.

Victron macht auch keine Aussagen, wann der neue Service für die Solar Charger zur Verfügung steht. Bis dahin käme man nur an die Daten, die die App in der Übersicht aller Geräte anzeigt. Diese werden im Device Advertising übertragen. Sie sind zwar verschlüsselt, aber Victron hat erklärt, wie man den Schlüssel auslesen und die Daten entschlüsseln kann.

**Comment by helgeerbe** [comment]
So, habe eine erste Victron BLE Lib-implemtiert und kann die Daten jetzt vom Smart Solar Charger empfangen, und entschlüsseln.

````log
16:29:26.472 > Advertised Device: Name: SmartSolar SERIALNUMBER, Address: e3:34:b8:98:26:8e, manufacturer data: e102100256a0013a06df8ed51f257a85162c3bf89ce9 
16:29:26.485 > Battery Current: 12 <- 0.1 A
16:29:26.487 > Battery Voltage 2617 <- 0.01 V
16:29:26.488 > Charger Error Code 00
16:29:26.490 > Device State 03
16:29:26.491 > Load Current: 511 <- 0.1 A
16:29:26.492 > PV Power: 34 <- 1W
16:29:26.494 > Yield Today: 137 <- 0.01 kWh
````

Daten werden so jede Sekunde aktualisiert.

**Reply by MrZwickie** [reply]
> So, habe eine erste Victron BLE Lib-implemtiert und kann die Daten jetzt vom Smart Solar Charger empfangen, und entschlüsseln.

Das klingt ja klasse: Sehe ich das richtig, dass wir dann evtl. tatsächlich mit einer Erweiterung zur Nutzung mehrerer SmartSolars an einer OpenDTU (mit _einer_ Batterie und _einem_ Wechselrichter) ohne serielles Kabel rechnen könnten? Genau sowas würde ich mir nämlich zur Erweiterung meines 600W-O/W-"Garagenkraftwerks" wünschen. :) 

Vielen Dank übrigens an Dich, @helgeerbe (und auch an tbnobody, falls er das liest) für die tollen Projekte!

**Reply by helgeerbe** [reply]
Siehe meinen Kommentar unten. Wenn Victron es zukünftig erlaubt, alle Parameter auszulesen, werde ich es viellicht als Compile-Option einbinden. Bis dahin führt kein Weg am Kabel vorbei.

**Reply by MrZwickie** [reply]
> Bis dahin führt kein Weg am Kabel vorbei. [...] D. h. ich werde jetzt erstmal das Victron Interface so erweitert, dass mehrere MPPT per Kabel angeschlossen werden können. 

Das macht nichts - wäre ja ebenfalls schon super! 👍

**Comment by helgeerbe** [comment]
Hmm, bin etwas ernüchtert:

![image](https://user-images.githubusercontent.com/59169507/234566072-2dddfc7d-e172-45da-8122-06330fc07d4d.png)

Die Einbindung der Bluetooth und AES Lib (für die Entschlüsselung) verbraucht fasst den gesamten Platz für den Sketch. Es ist kaum noch was übrig für zukünftige Erweiterungen.

Der Mehrwert ist aktuell sehr gering, da man nur einen sehr eingeschränkten Satz an Parametern auslesen kann. Ve.Direct über Kabel bietet da doch viel mehr. Außerdem kommt es häufiger zu TCP Abbrüchen. Bluetooth und Wifi teilen sich die selbe Antenne. Es wird zwar wieder automatisch verbunden, aber schön ist das nicht.

Ich stelle die Bluetooth Anbindung erst mal zurück. Zumindest solange bis Victron (wie angekündigt) den vollen Zugriff auf alle Parameter erlaubt. Wenn es jemanden interessiert, hier ist mein Test-Code: https://gist.github.com/helgeerbe/5b7bf40f9437bdeed22ced0e16989d50

D. h. ich werde jetzt erstmal das Victron Interface so erweitert, dass mehrere MPPT per Kabel angeschlossen werden können.

**Reply by swingstate** [reply]
Wenn Du weiter an Ve.Direct arbeitest, kannst Du [hier](https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/185#discussioncomment-5724513) mal einen Blick drauf werfen? Vielleicht lässt sich das verbinden im Sinne von doppelte Arbeit vermeiden.

**Reply by Tyberyus** [reply]
Guten Abend und erstmal vielen Dank für deine Arbeit ^^ ich gehe da mit dir mit. Solange nicht alles mit Bluetooth geht würde ich da auch keine Zeit weiter investieren ich war bloß erfreut als ich die libs gefunden hatte^^

**Reply by madmartin** [reply]
@helgeerbe Zum Thema Platz im Flash kann ich noch empfehlen mal bei Tasmota einen Blick auf das "safeboot partition schema" zu werfen. Dort wird in eine kleine Partition ein "safeboot" image gelegt, über das man immer die Haupt-Firmware in der (größeren) Flash-Partition updaten kann (wenn mal was schiefläuft).  https://tasmota.github.io/docs/Safeboot/ Dort kann die Haupt-Firmware bis zu 2880kb groß sein!

**Reply by madmartin** [reply]
> 
> D. h. ich werde jetzt erstmal das Victron Interface so erweitert, dass mehrere MPPT per Kabel angeschlossen werden können.

@helgeerbe 

Da möchte ich gerne mal nach dem Stand fragen... ich bekomme nämlich in Kürze einen zweiten SmartSolar MPPT Lader und würde dann auch mit experimentieren und programmieren anfangen.

> UARTS sind eigentlich keine mehr frei. 

Nach meinen Recherchen müsste der dritte UART noch frei sein, denn das TWAI (Canbus) Interface scheint nicht einen UART zu belegen. In den original-Dokumenten  https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/uart.html und https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/twai.html ist nichts zu lesen daß es zwischen TWAI und UART Interface Abhängigkeiten gibt.

Schon mal mit Softserial was probiert? https://github.com/plerup/espsoftwareserial ?

**Reply by spcqike** [reply]
> @helgeerbe Zum Thema Platz im Flash kann ich noch empfehlen mal bei Tasmota einen Blick auf das "safeboot partition schema" zu werfen. Dort wird in eine kleine Partition ein "safeboot" image gelegt, über das man immer die Haupt-Firmware in der (größeren) Flash-Partition updaten kann (wenn mal was schiefläuft). https://tasmota.github.io/docs/Safeboot/ Dort kann die Haupt-Firmware bis zu 2880kb groß sein!

bin ich bei dir. aber wäre das nicht eher etwas für die "Haupt"-OpenDTU? Ich kann nicht abschätzen, wie umfangreich die Änderungen dafür wären oder wie kompliziert folgende Aktuallisierungen und Merges von OpenDTU zu OpenDTU-onBattery, sehe so eine grundlegende Änderung eher "im Ursprung" :)

**Reply by helgeerbe** [reply]
Ja, da bin ich bei @spcqike . Ich versuche möglichst keine Änderungen an der originalen openDTU durchzuführen.

**Reply by helgeerbe** [reply]
Stand habe ich noch nicht wirklich. 

Grundsätzlich würde ich:
- Softserial nutzen (die Performance scheint gut genug zu sein)
- Vedirect als Array auslegen, und gleich um die SmartShunt Attribute erweitern
- In jedem loop Durchgang genau ein VeDirect Element lesen, aber auch die Pollrate berücksichtigen.
- 2 Charger und ein Shunt sollte von der Performance noch funktionieren.

**Comment by helgeerbe** [comment]
No um meine Idee noch festzuhalten:

Wenn Victron vollständigen Zugriff ermöglicht, könnte man einen zweiten ESP die Bluetooth Kommunikation machen lassen und die Daten dann per Serial Interface an OpenDTU-onBattery übergeben. Wäre zumindest eine Möglichkeit.

**Reply by swingstate** [reply]
Die Idee hatte ich auch schon, wollte das hier aber nicht weiter verkomplizieren. 
Ein ESP könnte sich um den/die Inverter kümmern, ein zweiter um das Messen und bereitstellen von Daten von MPPTs, Shunts, AC Ladegeräten etc..

**Reply by Tyberyus** [reply]
Wäre eine Runde Lösung ^^ hätte ich Zeit könnte ich dazu ein Projekt starten. Habe meine Diplomarbeit über Bluetooth geschrieben und kenne mich da ziemlich gut aus ^^
